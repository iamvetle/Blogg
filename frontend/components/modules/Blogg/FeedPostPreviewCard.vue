<template>
	<div class="article grid-cols-12 grid gap-8 mb-16">

		<div id="left-col-article" class="col-start-1 col-end-9">
			
			<span class="flex items-center w-full mb-2">
				<span class="user-picture flex items-center">
					<slot name="profile_picture"></slot>
				</span>
				<span class="flex items-center justify-between w-full">

					<span class="flex items-center full_name">
						<slot name="author"></slot>

					</span>

					<slot name="date_published"></slot>

				</span>
			</span>
			<div class="mb-4 w-full">

				<slot name="title"></slot>


				<slot name="content"></slot>

			</div>

			<span class="flex items-center justify-between">
				<span class="flex items-center space-x-2">
					<slot name="lesmer"></slot>
				
				<span class="flex items-center">
					<slot name="tags"></slot>
				</span>
				</span>

				<span class="flex items-center">
					<span id="comments-count" class="me-2">
						<slot name="amount-of-comments"></slot>
					</span>

					<span id="save-article" class="me-2">
						<slot name="save-article-icon"></slot>
					</span>

					<span id="article-options" class="">

						<slot name="more-options-icon"></slot>

					</span>

				</span>
			</span>
		</div>

		<div id="right-col-article" class="col-span-4 flex items-center justify-center">
			<slot name="article_image">This is were the article image is.</slot>
		</div>
	</div>
</template>

<script setup lang="ts">

// temp solution


</script>

<style scoped></style>