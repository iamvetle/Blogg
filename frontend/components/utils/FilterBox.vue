<template>
    <div v-for="(option, index) in props.listOfOptions" :class="props.class" class="w-full">
        <span>
            <BaseCheckboxOption v-model="selected[option]" @update:model-value="updateList" :key="index" :label="option" v-bind="$attrs"/>
        </span>
    </div>
</template>

<script setup lang="ts">
/**
 * Filter box
 * 
 * @var props.listOfoptions - An array with strings should be passed as a prop to this component. 
 * 
 * Each item in the array will be displayed as a checkbox option.
 * 
 * @updateList - The event will return an array back with only the selected items, the boxes
 * that is checked.
 * 
 * I have to use it like this " ... v-model="selected[option]" @update:model-value="updateList" />"
 * so that I don't emit TWICE and then fetch twice
 * 
 */

/**
 * Stores the list of selected items
 * 
 * in a { "item": true/false } state
 */
const selected = ref<any>({})
const emit = defineEmits(['output'])

/**
 * Returns a computed propety array containing only the items that are selected
 * @returns A list of the checked checkboxes
 */
const selectedItems = computed(() => {
    let only_true_list = []

    for (let [key, value] of Object.entries(selected.value)) {
        if (value === true) {
            only_true_list.push(key)
        }
    }
    return only_true_list
})

const updateList = () => {
    emit('output', selectedItems)
}

/**
 * The prop contains an array of all of the options that can be selected from
 */
const props = defineProps<{
    listOfOptions: any,
    class?:string,
    inputClass?:string,
}>();

onMounted(() => {
  console.log('FilterBox component is mounted.');
});

onActivated(() => {
  console.log('FilterBox component is activated.');
});

onUnmounted(() => {
  console.log('FilterBox component is unmounted.');
});

onDeactivated(() => {
  console.log('FilterBox component is deactivated.');
});

</script>

<style lang="scss" scoped></style>