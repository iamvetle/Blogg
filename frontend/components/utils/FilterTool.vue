<template>
	<div class="py-1" v-if="postStore.allTags">
		<div class="py-1 ml-2" v-for="(category, index) in postStore.allTags.results">
			<BaseCheckboxOption :key="index" v-model="selectedCategories[category.name]" :label="category.name" />
		</div>		
	</div>
</template>

<script setup lang="ts">
import { usePostStore } from '~/store/postStore';
import { usePaginationStore } from '~/store/paginationStore';

/** 
 * The component is a base for filtering. It has input fields that can be 
 * customized. The component is susposed to be associated to a fetch function, or similar.
 * 
 * The component is NOT finished. An unfinished idea!  
 * 
 */

const baseGroundURL = "http://localhost:8888/api/feed/"
const setupTagURL = `${baseGroundURL}?`

const selectedCategories = ref<any>({})

const postStore = usePostStore()
const paginationStore = usePaginationStore()

const selectedCategoryNames = computed(() => {
	return Object.keys(selectedCategories.value).filter(key => selectedCategories.value[key]);
});

/**
 * The computed function returns the custom url I would use in my api data fetch 
 * call for tag filtering
 * 
 * @returns - custom url ready for GET fetch   
 */
const customURL = computed(() => {
	let url = setupTagURL

	for (let category of selectedCategoryNames.value) {
		url += `categories=${category}&`
	}
	if (url != setupTagURL) {
		return url.slice(0, -1)
	} else {
		return baseGroundURL
	}
})


/**
 * Fetches new posts automatically based on the customURL
 */
	watchEffect(async () => {
		paginationStore.activeFetchURL = customURL.value 
		await getPostMultipleSnippet() // maybe I should please <- this one longer up?
	})

</script>



<style scoped></style>
