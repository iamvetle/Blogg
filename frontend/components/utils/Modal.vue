<template>
    <div>
        <button type="button" id="open" @click="setOpen">Open</button>

            <Dialog :open="isOpen" @close="setClose">
                <div class="fixed w-full h-full flex items-center justify-center inset-0">

                    
                <div class="w-[500px] h-[250px] bg-primary text-onPrimary p-8">
                    <DialogPanel>
                        <div>
                            <DialogTitle class="text-center text-lg">Are you ready to publish the post?</DialogTitle>
                            <p>
                            </p>
                            <div class="mt-16">
                                <span id="choices" class="flex justify-center space-x-24">
                                    <span class="p-4 bg-onPrimary text-primary"><button id="close" @click="setClose" >Yes</button></span>
                                    <span class="p-4 bg-onPrimary text-primary"><button @click="setClose">No</button></span>
                                </span>
                            </div>
                        </div>
                    
                    </DialogPanel>
                </div>
            </div>

            </Dialog>
        </div>
</template>

<script setup lang="ts">
import {
    Dialog,
    DialogPanel,
    DialogTitle,
    // DialogDescription,
} from '@headlessui/vue'

const emit = defineEmits()

const props = defineProps<{
    
}>();

const selectedBeforePublished = () => {
    emit("controlForPublish")
}

const isOpen = ref<boolean>(false)

const setOpen = () => {
    isOpen.value = true
}

const setClose = () => {
    isOpen.value = false
}


</script>

<style scoped></style>