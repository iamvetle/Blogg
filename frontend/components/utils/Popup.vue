<template>
    <div id="root" v-if="isOpen" class="relative">
        <teleport to="#modal-root">
            <div id="modal" class="fixed inset-0 flex items-center justify-center z-50">
                <p>hello there</p>
                <button @click="isOpen = false">Close</button>
            </div>
        </teleport>
    </div>
</template>

<script setup lang="ts">



const props = withDefaults(defineProps<{
    initial:boolean
}>(), {
    initial:false
});


const isOpen = ref(props.initial)

</script>

<style scoped>





</style>