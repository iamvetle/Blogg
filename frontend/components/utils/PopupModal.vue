<template>
    <div>
        <UModal v-model="isOpen" prevent-close>
            <UCard>
                <div data-test="inner_popup_modal" class="p-4">


                    <span class="text-center text-2xl text-onSurface">Do you want to publish the post?</span>
                    <p>
                    </p>
                    <div class="mt-4">
                        <span id="choices" class="flex justify-end space-x-6">

                            <span>
                                <button data-test="do_abort_option" class="p-1 text-primary
                                        
                                        " @click="doAbort">
                                    {{ abortText }}
                                </button>
                            </span>

                            <span class="">
                                <button data-test="do_confirm_option" id="cancel" class="p-1 text-primary
                                        " @click="doConfirm">
                                    {{ confirmText }}
                                </button>
                            </span>

                        </span>
                    </div>
                </div>


            </UCard>
        </UModal>
    </div>
</template>

<script setup lang="ts">
/**
 * I dont need to have a prop closing and opening it if I just have show hide outside instead
 */


const isOpen = ref(true)

const emit = defineEmits(["confirm", "abort"])

withDefaults(defineProps<{
    // isOpen: boolean;
    confirmText?: string;
    abortText?: string;
}>(), {
    confirmText: "Confirm",
    abortText: "Abort"

})

const doConfirm = () => {
    emit("confirm")
}

const doAbort = () => {
    emit("abort")
}


</script>