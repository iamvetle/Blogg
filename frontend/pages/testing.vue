<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>