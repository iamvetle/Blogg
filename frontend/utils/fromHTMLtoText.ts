//@ts-ignore
import {htmlToText} from 'html-to-text'

/**
 * Converts HTML string to plain text.
 * @param htmlString - The HTML string to convert.
 * @returns The plain text representation of the HTML string.
 */
export const fromHTMLtoText = (htmlString: string) => {
    // Creates a "fake" document (that I can use server side)

    return htmlToText(htmlString) || ""
};

