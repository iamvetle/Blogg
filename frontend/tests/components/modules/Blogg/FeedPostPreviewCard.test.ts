import FeedPostPreviewCard from '~/components/modules/Blogg/FeedPostPreviewCard.vue';
import { shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { testSnippetPostSingle } from '~/tests/testingHelpers/testConstants';
import { fromHTMLtoText } from '~/utils/fromHTMLtoText';

import { render, screen, within } from '@testing-library/vue'
import exampleImage from "~/assets/example-image.jpg";

let pinia = createTestingPinia();
let authStore: any;
let paginationStore: any;
// const feedPostsURL = "http://localhost:8888/api/posts/feed/"
// const feedPostsURLPreviousPage = `${feedPostsURL}?page=2`
// const feedPostsURLNextPage = `${feedPostsURL}?page=5`

let post = testSnippetPostSingle

let mockFromHTMLtoText = (value: any) => fromHTMLtoText(value);


describe("testing PostArticle", async () => {

    beforeEach(() => {
        // vi.stubGlobal("fromHTMLToText", () => {
        //     return null
        // })

        // vi.stubGlobal("fromHTMLToText", () => null);

        authStore = useAuthStore(pinia)
        paginationStore = usePaginationStore(pinia)

    })

    const wrapper = shallowMount(FeedPostPreviewCard, {
        global: {
            plugins: [pinia],
            components: {},
            mocks: {
                fromHTMLtoText: mockFromHTMLtoText
            },
            stubs: {},
        },
        props: {
            post: post
        },
    });

    test("Should exist", () => {
        console.log(wrapper.html())
        expect(wrapper.exists()).toBe(true)
    })
    test("The title should render", () => {
        expect(wrapper.text()).toContain(post.title)
    })
    test("The content(snippet) gets rendered", () => {
        const rawTextForTest = fromHTMLtoText(post.content_snippet)
        expect(wrapper.text()).toContain(rawTextForTest)

    })
    test("Should render the authors full name", async () => {
        // vi.stubGlobal("fromHTMLToText", () => null);
        render(FeedPostPreviewCard, {
            props: {
                post: post
            }, global: { mocks: { fromHTMLtoText: mockFromHTMLtoText } }
        })



        const authorPost = screen.getByTestId("post_author_name")
        within(authorPost).getByText(post.author.first_name, { exact: false })

    })
    test('should the "date published"', () => {
        const datePubText = wrapper.find("[data-test='post_date_published']")
        expect(datePubText.text()).toBe(post.date_published)

    })
    test('Should render "Les mer"', () => {
        expect(wrapper.html()).toContain("Les mer")
    })
    test('The image should have the correct temporary example image', () => {

        expect(wrapper.find("[data-test='post_article_image']").attributes("src")).toBe(exampleImage)
    })
    test("Should render the number of comments a post has", () => {

        expect(wrapper.html()).toContain(post.num_of_comments);
    })
    test('Should match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
})
