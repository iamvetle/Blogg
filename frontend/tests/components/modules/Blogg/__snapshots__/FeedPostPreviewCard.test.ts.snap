// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing PostArticle > Should match snapshot 1`] = `
"<div class=\\"article grid-cols-12 grid gap-8 mb-16\\">
  <div data-article=\\"left-col-article\\" class=\\"col-start-1 col-end-9\\"><span class=\\"flex items-center w-full mb-2\\"><span class=\\"user-picture flex items-center\\"><!-- User profile picture--><baseimage src=\\"/assets/account-pin-circle-line.svg\\" alt=\\"Profile picture\\" class=\\"w-6 h-auto me-2\\"></baseimage></span><!-- Author--><span class=\\"flex items-center justify-between w-full\\"><span class=\\"flex items-center full_name\\"><nuxtlink to=\\"/user/kingOfMario23\\" class=\\"cursor-pointer break-words\\"><span data-testid=\\"post_author_name\\" class=\\"font-bold\\">Timmy Bakke</span></nuxtlink></span><!-- Date published --><span data-test=\\"post_date_published\\" class=\\"cursor-default font-light\\">2023-23-2</span></span></span>
    <div class=\\"mb-4 w-full\\">
      <!--Title-->
      <nuxtlink to=\\"/post/14\\" class=\\"cursor-pointer break-words\\">
        <h3 class=\\"text-[28px] mb-2\\">Just testing</h3>
      </nuxtlink>
      <!--Content-->
      <div class=\\"break-words mb-2\\">SOMETHING THAT SHOULD BE PROCESSED

        Body content goes here</div>
    </div><span class=\\"flex items-center justify-between\\"><span class=\\"flex items-center space-x-2\\"><!-- Les mer --><span class=\\"cursor-pointer text text-primary hover:text-primaryFixed\\"> Les mer </span>
    <!--Tags--><span class=\\"flex items-center\\"><span class=\\"me-1\\"><basetag text=\\"Development\\"></basetag></span><span class=\\"me-1\\"><basetag text=\\"Music\\"></basetag></span></span></span><span class=\\"flex items-center\\"><!--Amount of comments--><span data-article=\\"comments-count\\" class=\\"me-2\\"><span class=\\"cursor-default space-x-1 flex items-center\\"><span>5</span>
    <icon name=\\"has-comments\\" class=\\"h-6\\"></icon></span></span>
    <!--Save article--><span data-article=\\"save-article\\" class=\\"me-2\\"><!--v-if--></span>
    <!--Article options--><span data-article=\\"article-options\\" class=\\"\\"><baseiconmoreoptions widthprop=\\"24\\" heightprop=\\"24\\" colorprop=\\"fill-black\\"></baseiconmoreoptions></span></span></span>
  </div>
  <div data-article=\\"right-col-article\\" class=\\"col-span-4 flex items-center justify-center\\">
    <!-- Article image --><img data-test=\\"post_article_image\\" src=\\"/assets/example-image.jpg\\" alt=\\"Bilde til artikkel\\" class=\\"w-full h-auto\\">
  </div>
</div>"
`;
